### PR TITLE
Add component-level memorization on custom color

### PR DESCRIPTION
**Pull request by samchan0221**
Ref #1320

### CODE REVIEW

File: portal/src/ThemeConfigurationWidget.tsx [hunk 1]
```diff
@@ -212,6 +212,49 @@ const ThemeConfigurationWidget: React.FC<ThemeConfigurationWidgetProps> =
...
+        const { isDarkTheme, ...colors } = lightTheme;
+        return colors;
+      }
+      const { isDarkTheme, ...colors } = DEFAULT_LIGHT_THEME;
+      return colors;
+    });
+
+    const onPickerChangePrimaryColor = useCallback(
+      (color: string) => {
+        setCustomTheme({ ...customTheme, primaryColor: color });
```
Comment (louischan-oursky) [inline portal/src/ThemeConfigurationWidget.tsx:236]: setState can accept an update function. By using the update function, we can avoid depending on the current state.

File: portal/src/ThemeConfigurationWidget.tsx [hunk 2]
```diff
@@ -212,6 +212,49 @@ const ThemeConfigurationWidget: React.FC<ThemeConfigurationWidgetProps> =
...
+        onChangePrimaryColor(color);
+      },
+      [customTheme, onChangePrimaryColor]
+    );
+
+    const onPickerChangeTextColor = useCallback(
+      (color: string) => {
+        setCustomTheme({ ...customTheme, textColor: color });
```
Comment (louischan-oursky) [inline portal/src/ThemeConfigurationWidget.tsx:244]: setState uses update function.

File: portal/src/ThemeConfigurationWidget.tsx [hunk 3]
```diff
@@ -212,6 +212,49 @@ const ThemeConfigurationWidget: React.FC<ThemeConfigurationWidgetProps> =
...
+        onChangeTextColor(color);
+      },
+      [customTheme, onChangeTextColor]
+    );
+
+    const onPickerChangeBackgroundColor = useCallback(
+      (color: string) => {
+        setCustomTheme({ ...customTheme, backgroundColor: color });
```
Comment (louischan-oursky) [inline portal/src/ThemeConfigurationWidget.tsx:252]: setState uses update function.

**Review by louischan-oursky — COMMENTED** [review]
Memoize the custom theme inside ThemeConfigurationWidget has one caveat. The discard change button does not discard the custom theme :( But fixing that caveat is not trivial so let ignore that for now.

**Comment by samchan0221** [comment]
> Memoize the custom theme inside ThemeConfigurationWidget has one caveat. The discard change button does not discard the custom theme :( But fixing that caveat is not trivial so let ignore that for now.

I also thought about that. Passing state top-down or getting ref bottom-up are both not an elegant way to solve it. 
Maybe adding another useState in the configform hooks can do?

**Comment by louischan-oursky** [comment]
Build failed

**Pull request merged**